universe-instances-extended/src/Data/Universe/Instances/Extended.hs

```diff
@@ -28,10 +42,46 @@ instance (Representable f, Finite (Rep f), Ord (Rep f), Universe a)
 instance (Representable f, Finite s, Ord s, Finite (Rep f), Ord (Rep f), Universe a)
   => Universe (TracedT s f a)
   where universe = map tabulate universe
+instance (Finite a, Ord a)
+  => Universe (Equivalence a)
+  where universe = [Equivalence ((==) `on` (`M.lookup` m)) | m <- partitions universeF]
+          where partitionA :: Applicative f => (a -> f Bool) -> [a] -> f ([a], [a])
+                partitionA f xs = mconcat <$> traverse inject xs
+                            where inject x = (\b -> ([x | b], [x | not b])) <$> f x
+                partitions :: Ord a => [a] -> [Map a Natural]
+                partitions = go 0
+                       where go _ []       = [M.empty]
+                             go n (x : xs) = do
+                               (eq, notEq) <- partitionA (const [False, True]) xs
+                               rest        <- go (n + 1) notEq
+                               pure $ M.fromList [(x', n) | x' <- x : eq] `M.union` rest
+instance (Universe a, Finite b, Ord b)
+  => Universe (Op a b) where
+#if MIN_VERSION_base(4,7,0)
+        universe = coerce (universe :: [b -> a])
+#else
+        universe = map Op universe
+#endif
+instance (Finite a, Ord a)
+  => Universe (Predicate a)
+  where universe = map (Predicate . flip S.member) universe
 
 instance (Representable f, Finite (Rep f), Ord (Rep f), Finite a)
   => Finite (Co f a)
   where universeF = map tabulate universeF; cardinality = retag (cardinality :: Tagged (Rep (Co        f) -> a) Natural)
 instance (Representable f, Finite s, Ord s, Finite (Rep f), Ord (Rep f), Finite a)
   => Finite (TracedT s f a)
   where universeF = map tabulate universeF; cardinality = retag (cardinality :: Tagged (Rep (TracedT s f)) Natural)
+instance (Finite a, Ord a)
+  => Finite (Equivalence a)
+  where cardinality = retag (bell <$> (cardinality :: Tagged a Natural))
+              where -- Compute the Bell number for the given cardinality.
```

Comment:
what is *Bell number*, add a reference

Reply:
I like the wolfram.com definition.
```suggestion
              where -- Compute the Bell number for the given cardinality.
                    -- The Bell number is the "number of ways a set of n elements can be partitioned into nonempty subsets"
                    -- http://mathworld.wolfram.com/BellNumber.html
```
Do you think it is also worth stating somewhere the relation between equivalences and partitions? E.g.:
https://proofwiki.org/wiki/Fundamental_Theorem_on_Equivalence_Relations
https://proofwiki.org/wiki/Relation_Induced_by_Partition_is_Equivalence